File: functions/mail-extraction/handler.py
import os
import boto3
import json
from langchain.chains import LLMChain
from langchain.prompts import PromptTemplate
from langchain_openai import ChatOpenAI

PROMPT_TEMPLATE = """
다음 이메일 내용에서 아래의 6가지 속성에 해당하는 정보를 JSON 형식으로 추출하세요.
각 도시명을 아래 `City Code`로 매핑하여 일관성 있게 표시하세요. 만약 목록에 없는 도시는 "unknown"으로 표시하세요.

[List of attributes]
1. Weight:type(integer)
2. ContainerSize:type(integer)
3. DepartureDate:type(Date)
4. ArrivalDate:type(Date)
5. DepartureCity:type(string) - 도시명과 `City Code` 목록 참조
6. ArrivalCity:type(string) - 도시명과 `City Code` 목록 참조

도시명과 `City Code` 목록:
- New York (JFK)
- Los Angeles (LAX)
- San Francisco (SFO)
- Chicago (ORD)
- London (LHR)
- Paris (CDG)
- Tokyo (HND)
- Beijing (PEK)
- Sydney (SYD)
- Seoul (ICN)
- Dubai (DXB)
- Frankfurt (FRA)
- Singapore (SIN)
- Hong Kong (HKG)
- Toronto (YYZ)
- Johor Bahru, Malaysia (JHB)
- Incheon, South Korea (ICN)
- Busan, South Korea (PUS)
- Frankfurt, Germany (FRA)
- Shenzhen, China (SZX)
- Hanoi, Vietnam (HAN)

이메일 내용:
{email_content}

JSON 형식 데이터만 응답으로 제공해 주세요. 목록에 없는 도시는 "unknown"으로 표시해 주세요. 찾을 수 없는 속성은 빈 값을 넣어주세요.
"""

def create_extraction_chain():
    openai_api_key = os.environ['OPENAI_API_KEY']
    llm = ChatOpenAI(api_key=openai_api_key, model="gpt-4o-mini", temperature=0, max_retries=2)
    prompt = PromptTemplate(template=PROMPT_TEMPLATE, input_variables=["email_content"])
    return LLMChain(llm=llm, prompt=prompt)

def lambda_handler(event, context):
    sqs_client = boto3.client("sqs")
    extract_queue_url = os.environ["EXTRACT_QUEUE_URL"]

    print("Lambda function invoked with event:", event)  # Log 전체 이벤트

    for record in event["Records"]:
        # 이메일 내용을 포함하는 메시지 본문 가져오기
        message = json.loads(record["body"])
        email_content = message.get("email_content", "")
        object_key = message.get("key", "")
        sender = message.get("sender", "")

        print(f"Processing email content: {email_content[:100]}...")  # Log 이메일 본문 일부
        print(f"S3 object key: {object_key}")  # Log S3 객체 키
        print(f"Sender: {sender}")  # Log 발신자 정보

        # 추출 체인 생성 및 실행
        extraction_chain = create_extraction_chain()
        result = extraction_chain.run(email_content=email_content)
        
        print("LLM extraction result:", result)  # Log LLM의 추출 결과

        result_cleaned = result.strip("```json\n").strip("\n```")
        
        try:
            result_json = json.loads(result_cleaned)
        except json.JSONDecodeError:
            result_json = {
                "error": "Failed to parse LLM response to JSON",
                "llm_output": result
            }
            print("JSON decoding error:", result_json)  # Log JSON 디코딩 오류 발생 시

        # 추출된 정보와 S3 키, 발신자 정보 포함하여 SQS에 전송
        response = sqs_client.send_message(
            QueueUrl=extract_queue_url,
            MessageBody=json.dumps({
                "data": result_json,
                "key": object_key,
                "sender": sender
            })
        )
        print("SQS send message response:", response)  # Log SQS 전송 응답

    print("Lambda function execution completed.")
    return {
        "statusCode": 200,
        "body": result_json
    }

File: functions/mail-extraction/handler.js
const { S3Client, GetObjectCommand } = require('@aws-sdk/client-s3');
const { SQSClient, SendMessageCommand } = require('@aws-sdk/client-sqs');
const { TextractClient, AnalyzeDocumentCommand } = require('@aws-sdk/client-textract');
const { v4: uuidv4 } = require('uuid');
const { OpenAI } = require('openai');
const mammoth = require('mammoth');
const XLSX = require('xlsx');

const cityMapping = require('./cityMapping');
const PROMPT_TEMPLATE = require('./promptTemplate');

// AWS Clients Initialization
const s3Client = new S3Client({ region: 'ap-northeast-2' });
const sqsClient = new SQSClient({ region: 'ap-northeast-2' });
const textractClient = new TextractClient({ region: 'ap-northeast-2' });
const openAI = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

console.log('Loaded City Mapping:', cityMapping);
console.log('Loaded Prompt Template:', PROMPT_TEMPLATE);

// Validate and Set Defaults for Extracted Data
const validateExtractedData = (data) => {
  const defaultValues = {
    Weight: 0,
    ContainerSize: 0,
    DepartureDate: '',
    ArrivalDate: '',
    DepartureCity: 'unknown',
    ArrivalCity: 'unknown',
  };

  const validContainerSizes = [1, 2, 3, 4];
  const cityCodes = Object.values(cityMapping);

  return {
    Weight: typeof data.Weight === 'number' && data.Weight > 0 ? data.Weight : defaultValues.Weight,
    ContainerSize: validContainerSizes.includes(data.ContainerSize)
      ? data.ContainerSize
      : defaultValues.ContainerSize,
    DepartureDate:
      typeof data.DepartureDate === 'string' ? data.DepartureDate : defaultValues.DepartureDate,
    ArrivalDate:
      typeof data.ArrivalDate === 'string' ? data.ArrivalDate : defaultValues.ArrivalDate,
    DepartureCity: cityCodes.includes(data.DepartureCity) ? data.DepartureCity : 'unknown',
    ArrivalCity: cityCodes.includes(data.ArrivalCity) ? data.ArrivalCity : 'unknown',
  };
};

// Extract Text from Images using Textract
const extractTextFromImages = async (imageFiles) => {
  let combinedText = '';
  for (const image of imageFiles) {
    const params = {
      Document: { Bytes: image },
      FeatureTypes: ['TABLES', 'FORMS'],
    };
    const command = new AnalyzeDocumentCommand(params);
    const response = await textractClient.send(command);
    const lines = response.Blocks.filter((block) => block.BlockType === 'LINE');
    combinedText += lines.map((line) => line.Text).join(' ') + ' ';
  }
  return combinedText.trim();
};

// Extract Text from Excel Files
const extractTextFromExcel = (buffer) => {
  const workbook = XLSX.read(buffer, { type: 'buffer' });
  let extractedText = '';

  workbook.SheetNames.forEach((sheetName) => {
    const sheet = workbook.Sheets[sheetName];
    const jsonData = XLSX.utils.sheet_to_json(sheet, { header: 1 });

    jsonData.forEach((row) => {
      extractedText += row.join(' ') + '\n';
    });
  });

  return extractedText.trim();
};

// Extract Text from DOC/DOCX Files using Mammoth
const extractTextFromWord = async (buffer) => {
  try {
    const result = await mammoth.extractRawText({ buffer });
    return result.value;
  } catch (error) {
    console.error('Error extracting text from Word document:', error.message);
    return '';
  }
};

// Lambda Handler
exports.handler = async (event) => {
  for (const record of event.Records) {
    try {
      const message = JSON.parse(record.body);
      const { email_content, attachments, sender } = message;

      console.log(`Processing message: ${JSON.stringify(message)}`);

      const truncatedEmailContent = email_content || '';

      // Process Attachments
      const textFiles = [];
      const imageFiles = [];
      const unsupportedFiles = [];

      for (const key of attachments) {
        const params = { Bucket: 'mails-to-files', Key: key };
        const command = new GetObjectCommand(params);
        const data = await s3Client.send(command);
        const fileBuffer = await streamToBuffer(data.Body);
        const fileExtension = key.split('.').pop().toLowerCase();

        if (['txt', 'csv', 'docx', 'doc'].includes(fileExtension)) {
          let extractedText = '';
          if (['doc', 'docx'].includes(fileExtension)) {
            extractedText = await extractTextFromWord(fileBuffer);
          } else {
            extractedText = fileBuffer.toString('utf-8');
          }
          textFiles.push({ filename: key, file: Buffer.from(extractedText, 'utf-8') });
        } else if (['jpg', 'jpeg', 'png'].includes(fileExtension)) {
          imageFiles.push(fileBuffer);
        } else if (['xls', 'xlsx'].includes(fileExtension)) {
          try {
            const excelContent = extractTextFromExcel(fileBuffer);
            textFiles.push({ filename: key, file: Buffer.from(excelContent, 'utf-8') });
          } catch (error) {
            console.error(`Error processing Excel file (${key}):`, error.message);
            unsupportedFiles.push(key);
          }
        } else {
          unsupportedFiles.push(key);
        }
      }

      console.log(
        `Text files: ${textFiles.length}, Image files: ${imageFiles.length}, Unsupported files: ${unsupportedFiles.join(', ')}`,
      );

      let imageText = '';
      if (imageFiles.length > 0) {
        imageText = await extractTextFromImages(imageFiles);
      }

      const combinedContent = PROMPT_TEMPLATE.replace(
        '{email_content}',
        `${truncatedEmailContent}\nAttached file content:\n${textFiles
          .map((file) => file.file.toString('utf-8'))
          .join('\n')}\n${imageText}`,
      );

      console.log('Combined Content for OpenAI:', combinedContent);

      const llmResult = await openAI.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [
          {
            role: 'system',
            content: 'You are an assistant extracting attributes from email content.',
          },
          { role: 'user', content: combinedContent },
        ],
        max_tokens: 500,
      });
      console.log('OpenAI Response:', JSON.stringify(llmResult, null, 2));

      let extractedData = {};
      try {
        const content = llmResult.choices[0].message.content;
        const cleanedContent = content.replace(/```json\n?|```/g, '').trim();
        extractedData = JSON.parse(cleanedContent);
      } catch (error) {
        console.error('Failed to parse extracted data:', error.message);
        extractedData = {};
      }

      const validatedData = validateExtractedData(extractedData);

      const sqsMessage = {
        key: uuidv4(),
        sender,
        data: validatedData,
      };

      console.log('SQS Message Body:', JSON.stringify(sqsMessage, null, 2));

      const sendMessageCommand = new SendMessageCommand({
        QueueUrl: process.env.EXTRACT_QUEUE_URL,
        MessageBody: JSON.stringify(sqsMessage),
      });
      await sqsClient.send(sendMessageCommand);

      console.log('SQS Send Success');
    } catch (error) {
      console.error('Error processing record:', error.message);
    }
  }

  return { statusCode: 200, body: JSON.stringify({ message: 'Processing complete.' }) };
};

// Utility Functions
const streamToBuffer = async (stream) => {
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks);
};
